File: restapi/services/auth_service/ref/api/response.go
package api

import (
	"auth-service/internal/db"
	"time"
)

// ============================================================================
// Generic API Response
// ============================================================================

type APIResponse[T any] struct {
	Success bool   `json:"success"`
	Message string `json:"message,omitempty"`
	Data    T      `json:"data,omitempty"`
}

func SuccessResponse[T any](data T) APIResponse[T] {
	return APIResponse[T]{
		Success: true,
		Data:    data,
	}
}

func SuccessMessageResponse[T any](message string, data T) APIResponse[T] {
	return APIResponse[T]{
		Success: true,
		Message: message,
		Data:    data,
	}
}

func ErrorResponse(message string) APIResponse[any] {
	return APIResponse[any]{
		Success: false,
		Message: message,
	}
}

type userResponse struct {
	Username          string    `json:"username"`
	Email             string    `json:"email"`
	PasswordChangedAt time.Time `json:"password_changed_at"`
	CreatedAt         time.Time `json:"created_at"`
}

func newUserResponse(user db.User) userResponse {
	return userResponse{
		Username:          user.Username,
		Email:             user.Email,
		PasswordChangedAt: user.PasswordChangedAt.Time,
		CreatedAt:         user.CreatedAt.Time,
	}
}

type loginUserResponse struct {
	SessionID             string       `json:"session_id"`
	AcessToken            string       `json:"access_token"`
	AccessTokenExpiresAt  time.Time    `json:"access_token_expires_at"`
	RefreshToken          string       `json:"refresh_token"`
	RefreshTokenExpiresAt time.Time    `json:"refresh_token_expires_at"`
	User                  userResponse `json:"user"`
}

type renewAccessTokenResponse struct {
	AcessToken           string    `json:"access_token"`
	AccessTokenExpiresAt time.Time `json:"access_token_expires_at"`
}

File: restapi/services/auth_service/src/myapp/server/api/response.py
from datetime import datetime
from typing import Any, Generic, TypeVar

from pydantic import BaseModel

T = TypeVar("T")


class APIResponse(BaseModel, Generic[T]):
    success: bool
    message: str = ""
    data: T | None = None


def success_response(data: Any) -> APIResponse:
    return APIResponse(success=True, data=data)


def success_message_response(message: str, data: Any) -> APIResponse:
    return APIResponse(success=True, message=message, data=data)


def error_response(message: str) -> APIResponse:
    return APIResponse(success=False, message=message)


# ============================================================================
# Domain-specific response models
# ============================================================================

class UserResponse(BaseModel):
    username: str
    email: str
    password_changed_at: datetime | None = None
    created_at: datetime | None = None


class LoginUserResponse(BaseModel):
    session_id: str
    access_token: str
    access_token_expires_at: datetime
    refresh_token: str
    refresh_token_expires_at: datetime
    user: UserResponse


class RenewAccessTokenResponse(BaseModel):
    access_token: str
    access_token_expires_at: datetime
